eclim-debug.el

```diff
@@ -0,0 +1,118 @@
+  (let ((sourcepath (eclim--debug-sourcepath project)))
+    (format "jdb -attach %s -sourcepath%s "
+            port
+            sourcepath)))
+
+(defun eclim--debug-jdb-run-command (project args)
+  (let ((sourcepath (eclim--debug-sourcepath project))
+        (classpath (eclim/java-classpath project)))
+    (concat (format "jdb -sourcepath%s -classpath%s "
+                    sourcepath
+                    classpath)
+            (s-join " " args))))
+
+(defun eclim--debug-sourcepath (project)
+  (let ((projects (-snoc (eclim-project-dependencies project) project)))
+    (s-join ":" (-mapcat 'eclim--debug-project-sourcepath projects))))
+
+(defun eclim--debug-project-dir (project)
+  (cdr (assoc 'path (eclim/project-info project))))
+
+(defun eclim--debug-project-sourcepath (project)
+  (eclim--debug-read-sourcepath
+   (concat (file-name-as-directory (eclim--debug-project-dir project))
+           ".classpath")))
+
+(defun eclim--debug-read-sourcepath (classpath-file)
+  (let* ((root (car (xml-parse-file classpath-file)))
+         (classpathentries (xml-get-children root 'classpathentry))
+         (srcs (-filter 'eclim--debug-kind-src? classpathentries))
+         (paths-relative (-map 'eclim--debug-get-path srcs))
+         (paths-absolute (--map (concat (file-name-directory classpath-file) it) paths-relative)))
+    paths-absolute))
+
+(defun eclim--debug-kind-src? (classpathentry)
+  (let* ((attrs (xml-node-attributes classpathentry))
+         (kind (cdr (assq 'kind attrs))))
+    (string-equal kind "src")))
+
+(defun eclim--debug-get-path (classpathentry)
+  (let* ((attrs (xml-node-attributes classpathentry))
+         (path (cdr (assq 'path attrs))))
+    path))
+
+(defun eclim--debug-attach-to-mvn-when-ready (txt project)
+  (when (s-contains? "at address: 5005" txt)
+    (remove-hook 'comint-output-filter-functions
+                 'eclim--debug-attach-to-mvn-when-ready
+                 t)
+    (eclim-debug-attach 5005 project)))
+
+(defun eclim-debug-junit ()
+  (let ((project (eclim-project-name))
+        (classes (list "org.junit.runner.JUnitCore"
+                       (eclim-package-and-class))))
+    (jdb (eclim--debug-jdb-run-command project classes))))
+
+(defun eclim-debug-test-mvn ()
+  (interactive)
+  (let ((project (eclim-project-name)))
+    (with-current-buffer (eclim-maven-run-debug (concat "clean test -Dtest=" (file-name-base)))
+      (setq-local comint-prompt-read-only t)
+      (make-local-variable 'comint-output-filter-functions)
+      (add-hook 'comint-output-filter-functions
+                (lambda (txt) (eclim--debug-attach-to-mvn-when-ready txt project))))))
+
+(defun eclim-debug-attach (port project)
+  (interactive (list (read-number "Port: " 5005) (eclim-project-name)))
+  (jdb (eclim--debug-jdb-attach-command project port)))
+
+(defun eclim-debug-test ()
+  (interactive)
+  (cond ((eclim-java-junit-buffer?) (eclim-debug-junit))
+        ((eclim-java-testng-buffer?) (eclim-debug-test-mvn))
```

Comment:
I wrongly assumed that one will have `mvn` project. I need to fix that and make it work for ant as well... I would probably need help with ant here